### PR TITLE
fix clang tidy warning, add nullptr check

### PR DESCRIPTION
**Pull request by suzyw-w**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by vdimir** [comment]
What error exactly is about? We have `NOLINTNEXTLINE` below. Also clang-tidy build in out CI is ok.

**Review by vdimir — COMMENTED** [review]
What error exactly is about? We have `NOLINTNEXTLINE` below. Also clang-tidy build in out CI is ok.

**Comment by alexey-milovidov** [comment]
@SuzyWangIBMer you should also add this to `.clang-tidy` file in our source root to ensure it is covered by CI.

**Comment by alexey-milovidov** [comment]
I'm not sure that this change is correct.
If a function does not expect nullptr, you should pass by reference, not by pointer.

**Comment by SuzyWangIBMer** [comment]
> What error exactly is about? We have `NOLINTNEXTLINE` below. Also clang-tidy build in out CI is ok.

The error is about no check for ast nullptr, checking select nullptr does not guarantee ast to be valid. 
image

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Replace the calling code:

parsed_query.get() -> *parsed_query

**Comment by alexey-milovidov** [comment]
You should add this check to `.clang-tidy` file, otherwise it does not make sense.

**Comment by SuzyWangIBMer** [comment]
> You should add this check to `.clang-tidy` file, otherwise it does not make sense.
Since dereference of a null pointer is already in clang-tidy, hence just removed '//NOLINTNEXTLINE'

**Comment by SuzyWangIBMer** [comment]
> You should add this check to `.clang-tidy` file, otherwise it does not make sense.
Since dereference of a null pointer is already in clang-tidy, hence just removed '//NOLINTNEXTLINE'

**Comment by alexey-milovidov** [comment]
But now the code does not build.

**Comment by alexey-milovidov** [comment]
@SuzyWangIBMer The idea is the following:
1. Use references instead of pointers if possible.
2. After replacing pointers to references you can check for nullptr earlier and in less number of places.

**Comment by SuzyWangIBMer** [comment]
@alexey-milovidov  We do hold concern about this change. If we make it reference instead of pointer, then dynamic_cast for reference throws exception if it is not ASTSelectQuery, catching exception could become more expensive than the current design. Any suggestion you have based on this?

**Comment by alexey-milovidov** [comment]
Take an address for dynamic_cast.

File: programs/client/Client.cpp [hunk 1]
```diff
@@ -689,10 +689,9 @@ static bool queryHasWithClause(const IAST * ast)
     // defensive and avoid weird unexpected errors.
     // clang-tidy is confused by this function: it thinks that if `select` is
     // nullptr, `ast` is also nullptr, and complains about nullptr dereference.
```
Comment (vdimir) [inline programs/client/Client.cpp:691]:
```suggestion
```

**Review by vdimir — APPROVED** [review]

**Review by alexey-milovidov — APPROVED** [review]
LGTM

**Comment by alexey-milovidov** [comment]
Integration tests (asan, actions) [2/3] — fail: 1, passed: 615, flaky: 1
[Details](https://s3.amazonaws.com/clickhouse-test-reports/35540/5d2bd5b0c01ffbd60c79190e1ec941e3b3b8c4b9/integration_tests__asan__actions__[2/3].html)
@robot-clickhouse
Integration tests (release, actions) [1/2] — fail: 1, passed: 931, flaky: 0
[Details](https://s3.amazonaws.com/clickhouse-test-reports/35540/5d2bd5b0c01ffbd60c79190e1ec941e3b3b8c4b9/integration_tests__release__actions__[1/2].html)
@robot-clickhouse
Integration tests (thread, actions) [3/4] — fail: 3, passed: 426, flaky: 0
[Details](https://s3.amazonaws.com/clickhouse-test-reports/35540/5d2bd5b0c01ffbd60c79190e1ec941e3b3b8c4b9/integration_tests__thread__actions__[3/4].html)
@robot-ch-test-poll1
Stateful tests (thread, actions) — Timeout, fail: 1, passed: 115, skipped: 2
[Details](https://s3.amazonaws.com/clickhouse-test-reports/35540/5d2bd5b0c01ffbd60c79190e1ec941e3b3b8c4b9/stateful_tests__thread__actions_.html)

These tests have to be fixed before merging.

**Comment by alexey-milovidov** [comment]
Waiting for #36361.

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Comment by alexey-milovidov** [comment]
@SuzyWangIBMer I cannot merge your branch with master, you have to do it by yourself.

**Comment by SuzyWangIBMer** [comment]
> @SuzyWangIBMer I cannot merge your branch with master, you have to do it by yourself.
I am going to close this PR and open a new one.

**Pull request merged**